common/src/Microsoft.Azure.IIoT.AspNetCore/src/Auth/Extensions/DataProtectionBuilderEx.cs

```diff
@@ -0,0 +1,218 @@
+            if (client == null) {
+                throw new UnauthorizedAccessException("Cannot access keyvault");
+            }
+            var identifier = $"{config.KeyVaultBaseUrl.TrimEnd('/')}/keys/{keyName}";
+            return builder.ProtectKeysWithAzureKeyVault(client, identifier);
+        }
+
+        /// <summary>
+        /// Add blob key storage
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <param name="configuration"></param>
+        public static IDataProtectionBuilder AddAzureBlobKeyStorage(
+            this IDataProtectionBuilder builder, IConfiguration configuration = null) {
+            if (configuration == null) {
+                configuration = builder.Services.BuildServiceProvider()
+                    .GetRequiredService<IConfiguration>();
+            }
+
+            var containerName = "dataprotection";
+            var storage = new DataProtectionConfig(configuration);
+            if (string.IsNullOrEmpty(storage.BlobStorageConnString)) {
+                return builder;
+            }
+
+            var storageAccount = CloudStorageAccount.Parse(storage.BlobStorageConnString);
+            var relativePath = $"{containerName}/keys.xml";
+            var uriBuilder = new UriBuilder(storageAccount.BlobEndpoint);
+            uriBuilder.Path = uriBuilder.Path.TrimEnd('/') + "/" + relativePath.TrimStart('/');
+            var block = new CloudBlockBlob(uriBuilder.Uri, storageAccount.Credentials);
+            Try.Op(() => block.Container.Create());
+            return builder.PersistKeysToAzureBlobStorage(block);
+        }
+
+        /// <summary>
+        /// Try create new keyvault client using provided configuration. Will
+        /// try several combinations including managed service identity and
+        /// if allowed, visual studio tooling access.
+        /// </summary>
+        /// <param name="vaultUri"></param>
+        /// <param name="client"></param>
+        /// <param name="keyName"></param>
+        /// <returns></returns>
+        private static async Task<KeyVaultClient> TryKeyVaultClientAsync(string vaultUri,
+            IClientConfig client, string keyName) {
+            KeyVaultClient keyVault;
+
+            // Try reading with app and secret if available.
+            if (!string.IsNullOrEmpty(client.AppId) &&
+                !string.IsNullOrEmpty(client.AppSecret) &&
+                !string.IsNullOrEmpty(client.TenantId)) {
+                keyVault = await TryInititalizeKeyAsync("Application",
+                    vaultUri, keyName, $"RunAs=App; AppId={client.AppId}; " +
+                        $"AppKey={client.AppSecret}; TenantId={client.TenantId}");
+                if (keyVault != null) {
+                    return keyVault;
+                }
+            }
+
+            // Try using aims
+            keyVault = TryInititalizeKeyAsync("Managed Service Identity",
+                vaultUri, keyName).Result;
```

Comment:
again why not await?

Reply:
Great catch.   This should be async and only the extension shoud call it sync.